TombEngine/Objects/TR4/Trap/tr4_chain.cpp

```diff
@@ -1,29 +1,145 @@
+	{
+		auto& item = g_Level.Items[itemNumber];
+
+		auto* spark = GetFreeParticle();
+		spark->on = 1;
+		spark->sR = (GetRandomControl() & 0x1F) + 48;
+		spark->sG = spark->sR >> 1;
+		spark->sB = 0;
+		spark->dR = (GetRandomControl() & 0x3F) + 192;
+		spark->dG = (GetRandomControl() & 0x3F) + 128;
+		spark->dB = 32;
+		spark->colFadeSpeed = (GetRandomControl() & 3) + 12;
+		spark->fadeToBlack = 8;
+
+		spark->extras = 0;
+		spark->life = Random::GenerateInt(1,15);
+		spark->sLife = spark->life;
+		spark->dynamic = 1;
+
+		spark->xVel = (GetRandomControl() & 0x3F) - 32;
+		spark->yVel = -16 - (GetRandomControl() & 0xF);
+		spark->zVel = (GetRandomControl() & 0x3F) - 32;
+		spark->friction = 4;
+		spark->flags = SP_SCALE | SP_DEF | SP_ROTATE | SP_EXPDEF | SP_ITEM ;
+
+		if (GetRandomControl() & 1)
+		{
+			spark->flags |= SP_ROTATE;
+			spark->rotAng = GetRandomControl() & 0xFFF;
+			spark->rotAdd = (GetRandomControl() & 0x1F) - 16;
+		}
+
+		spark->x = pos.x + (GetRandomControl() & 0x1F) - 16;
+		spark->y = pos.y;
+		spark->z = pos.z + (GetRandomControl() & 0x1F) - 16;
+
+		spark->gravity = -16 - (GetRandomControl() & 0x1F);
+		spark->maxYvel = -16 - (GetRandomControl() & 7);
+		spark->scalar = spark->life < 32 ? 4 : 3;
+		spark->size = (GetRandomControl() & 7) + 20;
+		spark->sSize = spark->size;
+		spark->dSize = static_cast<int>(spark->size) >> 2;
+
+		spark->SpriteSeqID = ID_DEFAULT_SPRITES;
+		spark->SpriteID = 0;
+	}
+
+	void TriggerPendulumSpark(const GameVector& pos, const EulerAngles& angle, float length, int count)
+	{
+		for (int i = 0; i < count; i++)
+		{
+			auto& s = GetFreeSparkParticle();
+			s = {};
+			s.age = 1;
+			s.life = Random::GenerateFloat(5, 8);
+			s.friction = 0.05f;
+			s.gravity = 0.1f;
+			s.height = length + Random::GenerateFloat(16.0f, 22.0f);
+			s.width = length;
+			s.room = pos.RoomNumber;
+			s.pos = Vector3(pos.x + Random::GenerateFloat(-16, 16), pos.y + Random::GenerateFloat(6, 60), pos.z + Random::GenerateFloat(-16, 16));
+			float ang = TO_RAD(angle.y);
+			float vAng = -TO_RAD(angle.x);
+			Vector3 v = Vector3(sin(ang), vAng + Random::GenerateFloat(-PI / 16, PI / 16), cos(ang));
+			v.Normalize(v);
+			s.velocity =  v* Random::GenerateFloat(32, 64);
+			s.sourceColor = Vector4(1, 0.7f, 0.4f, 1);
+			s.destinationColor = Vector4(0.4f, 0.1f, 0, 0.5f);
+			s.active = true;
+		}
+	}    
+
 	void ControlChain(short itemNumber)
 	{
 		auto& item = g_Level.Items[itemNumber];
 
-		if (item.TriggerFlags)
+		if (!TriggerActive(&item))
+			return;
+
+		if (item.TriggerFlags < 0)
 		{
 			item.ItemFlags[2] = 1;
 			item.ItemFlags[3] = 75;
+			item.ItemFlags[5] = 1; // Set the item on fire when collide.
 
 			if (TriggerActive(&item))
 			{
 				*(int*)&item.ItemFlags[0] = 0x787E;
 				AnimateItem(&item);
+
+				auto pos = GetJointPosition(item, 4, Vector3i(0, 260, 0));
+				auto angle = GetBoneOrientation(item, 5);
+				byte r = 51 - ((GetRandomControl() / 16) & 6);
```

Comment:
Use `unsigned char` type and not `byte` type here, please.

Reply:
fixed